### PR TITLE
Presto 5436: Add support for using the AWS KMS for encryption keys for data in S3

### PR DESCRIPTION
**Pull request by natesammons-nasdaq**
Add support for using the AWS KMS for managing encryption keys for client-side encryption of data in S3.  This change adds a new configuration setting:

presto.s3.kms-key-id=KEY_UUID

This will cause PrestoS3FileSystem to use the AWS KMS for an encryption materials provider.  The existing setting, presto.s3.encryption-materials-provider, cannot be used with the AWS KMS because the KMS encryption materials provider requires a key ID passed into the constructor and can't initialize itself from Hadoop configuration files.

I've also added documentation on the S3 settings

### CODE REVIEW

File: presto-docs/src/main/sphinx/connector/hive.rst [hunk 1]
```diff
@@ -123,8 +123,6 @@ Property Name                                      Description
 
 ``hive.max-partitions-per-writers``                Maximum number of partitions per writer.                     100
 
-``hive.s3.sse.enabled``                            Enable S3 server-side encryption.                            ``false``
```
Comment (electrum) [inline presto-docs/src/main/sphinx/connector/hive.rst]: Did you mean to remove this?

Comment (electrum) [inline presto-docs/src/main/sphinx/connector/hive.rst]: Nevermind, I see this is moved above

File: presto-docs/src/main/sphinx/connector/hive-security.rst [hunk 2]
```diff
@@ -382,3 +382,119 @@ node.
...
+=============================
+
+The Hive Connector can read data from files stored in S3, assuming the Hive
+Metastore has tables defined as "external" which have locations defined as
+S3 URIs.
+
+Presto registers its own S3 filesystem for the following URI schemes:
+``s3://``, ``s3a://``, ``s3n://`` and it registers the default Hadoop S3
+FileSystem (``org.apache.hadoop.fs.s3.S3FileSystem``) for ``s3bfs://`` URIs.
+The following assumes your are using the native Presto S3 FileSystem implementation.
```
Comment (electrum) [inline presto-docs/src/main/sphinx/connector/hive-security.rst]: Typo: "you are" not "your are"

File: presto-docs/src/main/sphinx/connector/hive-security.rst [hunk 3]
```diff
@@ -382,3 +382,119 @@ node.
...
+
+Configuration Properties
+------------------------
+
+============================================ ====================================================== ==========
+Property Name                                Description                                            Default
+============================================ ====================================================== ==========
+``presto.s3.access-key``                     Default API access key to use
```
Comment (electrum) [inline presto-docs/src/main/sphinx/connector/hive-security.rst]: All of these properties should start with `hive.s3`. The `presto.s3` names are Hadoop configuration names which are internal to `PrestoS3FileSystem`. Users of the Hive connector should set the `hive.s3` properties in the Hive catalog properties file. The Hive connector will then set the `PrestoS3FileSystem` properties as appropriate.

Comment (natesammons-nasdaq) [inline presto-docs/src/main/sphinx/connector/hive-security.rst]: Ah, thanks -- I always forget about that translation layer from HiveClientConfig.

However, for "presto.s3.encryption-materials-provider" this is contained entirely in PrestoS3FileSystem - should the same be true for presto.s3.kms-key-id ?

File: presto-docs/src/main/sphinx/connector/hive-security.rst (hunk 2 above)
Comment (natesammons-nasdaq) [inline presto-docs/src/main/sphinx/connector/hive-security.rst]: I'll fix this

**Comment by electrum** [comment]
We didn't originally add the Presto config for `s3.encryption-materials-provider` since it was considered a "backdoor" for advanced users. Requiring users to drop a JAR file a directory and configure a class name is generally bad practice and not something we want to support. Unfortunately, there doesn't seem to be a good alternative.

Also, it seemed that anyone using this would need to have custom Hadoop config files specified, so specifying this as a Hadoop config property was fine. Is that a reasonable assumption? Or do people write providers that don't need any additional configuration (and thus the Hive catalog property file is enough)?

**Comment by natesammons-nasdaq** [comment]
Makes sense -- I think generally if people are writing custom materials providers they will either initialize themselves from the Hadoop config, or use something like java system properties.  The EMP code we use here at Nasdaq can do either, since in some cases (Apache Zeppelin for instance) its easier to deal with system properties and for systems like Presto dealing with Hadoop config files is easier...

File: presto-docs/src/main/sphinx/connector/hive-security.rst [hunk 4]
```diff
@@ -382,3 +382,119 @@ node.
...
+The Hive Connector can read data from files stored in S3, assuming the Hive
+Metastore has tables defined as "external" which have locations defined as
+S3 URIs.
+
+Presto registers its own S3 filesystem for the following URI schemes:
+``s3://``, ``s3a://``, ``s3n://`` and it registers the default Hadoop S3
+FileSystem (``org.apache.hadoop.fs.s3.S3FileSystem``) for ``s3bfs://`` URIs.
+The following assumes you are using the native Presto S3 FileSystem implementation.
+
+Configuration Properties
```
Comment (nezihyigitbasi) [inline presto-docs/src/main/sphinx/connector/hive-security.rst]: Some of the properties here and most of the properties below in the tuning section are not about security, so I am not sure whether we should put all these in the `hive-security` documentation. Maybe we should add these to the hive connector documentation, what do you think @electrum?

Comment (natesammons-nasdaq) [inline presto-docs/src/main/sphinx/connector/hive-security.rst]: Yeah, I actually think this should just be in a separate document, hive-s3.rst or something.  I feel like covering all the AWS/S3 related items in the same document makes sense.

Thoughts?

Comment (electrum) [inline presto-docs/src/main/sphinx/connector/hive-security.rst]: Yeah, let's move all this to an S3 section in the main Hive docs.

I don't think it's enough content yet to warrant a separate page.

**Comment by natesammons-nasdaq** [comment]
Moved docs into the main Hive section

File: presto-docs/src/main/sphinx/connector/hive.rst [hunk 5]
```diff
@@ -193,3 +193,119 @@ Hive Connector Limitations
...
+The following assumes you are using the native Presto S3 FileSystem implementation.
+
+S3 Configuration Properties
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+============================================ ====================================================== ==========
+Property Name                                Description                                            Default
+============================================ ====================================================== ==========
+``hive.s3.use-instance-credentials``         Use the EC2 metadata service to retrieve API           ``true``
+                                             credentials.  This works with IAM roles in EC2.
```
Comment (nezihyigitbasi) [inline presto-docs/src/main/sphinx/connector/hive.rst]: double space after `.`

**Comment by natesammons-nasdaq** [comment]
Any word on this PR?  Never heard back...

**Comment by natesammons-nasdaq** [comment]
I merged the changes since this branch was last touched.  Only merges were in the hive docs.

**Comment by electrum** [comment]
Merged, thanks!

**Comment by natesammons-nasdaq** [comment]
@electrum Would you like me to make the above changes?  Very happy to do it, just not sure given this PR is now merged and closed.

**Comment by natesammons-nasdaq** [comment]
Nevermind, looks like you already made those changes.  Thanks for merging this!

**Pull request closed without merging**